feat: add YOLOv8 as new model to the pipeline

francis-tanyx opened the pull request:
## PR Checklist
<!-- Please check if your PR fulfills the following requirements: -->

- [ ] Added label to the Pull Request for easier discoverability and search
- [ ] Commit Message meets guidelines as indicated in the URL https://github.com/intel-retail/automated-self-checkout/blob/main/CONTRIBUTING.md
- [ ] Every commit is a single defect fix and does not mix feature addition or changes
- [ ] Unit Tests have been added for new changes
- [ ] Updated Documentation as relevant to the changes
- [ ] All commented code has been removed
- [ ] If you've added a dependency, you've ensured license is compatible with repository license and clearly outlined the added dependency.
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide details below)

## What are you changing?
<!-- Please provide a short description of the updates that are in the PR -->

This PR will add  a new YOLOv8 model to the existing pipeline. YOLOv5 original model is still retained as per requested.
To change the model you will need to add the `YOLO` parameter to the `make` command. e.g. `YOLO=yolov8 make run-demo`.

## Issue this PR will close

close: #**666**

## Anything the reviewer should know when reviewing this PR?

We have only tested and run the code on the `make run-demo` and `make run-smoke-tests`. We also tried it on `make benchmark` but the igst.log did not return any output.

Docker composed is retained but we had to pull in the YOLOv8 sh files from loss-prevention reop.

## Test Instructions if applicable
<!-- How can the reviewers test your change? -->

We followed the steps on the Getting-Started Page to verify if the whole pipeline is this working (https://intel-retail.github.io/documentation/use-cases/automated-self-checkout/getting_started.html#step-by-step-instructions)

We also ran `make YOLO=yolov8 run-smoke-tests` 

## If the there are associated PRs in other repositories, please link them here (i.e. intel-retail/automated-self-checkout )

ejlee3 reviewed (changes requested):
When running `make run-demo` I get an error - permission denied in the dlstreamer container src-OvmsClientGst-1 `/script/entrypoint.sh: line 137: /home/pipeline-server/pipelines/yolov8s_roi.sh: Permission denied`

antoniomtz commented on src/pipelines/yolov8s_roi.sh:
```diff
@@ -0,0 +1,42 @@
+echo "decode type $DECODE"
+echo "Run YOLOv8 pipeline with ROI and Filtering on $DEVICE with batch size = $BATCH_SIZE"
+
+gstLaunchCmd="GST_DEBUG=1 GST_TRACERS=\"latency_tracer(flags=pipeline,interval=100)\" gst-launch-1.0 $inputsrc ! $DECODE ! gvaattachroi mode=1 file-path=/home/pipeline-server/pipelines/roi.json ! gvadetect batch-size=$BATCH_SIZE model-instance-id=odmodel name=detection model=models/object_detection/yolov8s/FP32/yolov8s.xml device=$DEVICE $PRE_PROCESS inference-region=1 object-class="$ROI" thresho ...
```
This is adding the same functionality as the loss-prevention use case. Can we make this script simpler, similar to the yolov5 pipeline. Only doing object detection and classification. No need to have ROI, or MQTT integration

francis-tanyx replied:
done

antoniomtz commented on src/pipelines/people_detection.sh:
```diff
@@ -0,0 +1,27 @@
+echo "decode type $DECODE"
+echo "Run person-detection pipeline on $DEVICE with batch size = $BATCH_SIZE"
+
+gstLaunchCmd="GST_DEBUG=\"GST_TRACER:7\" GST_TRACERS=\"latency_tracer(flags=pipeline,interval=100)\" gst-launch-1.0 $inputsrc ! $DECODE ! gvaattachroi roi=$ROI ! gvadetect batch-size=$BATCH_SIZE model-instance-id=odmodel name=detection model=models/object_detection/person-detection-0200/FP16-INT8/person-detection-0200.xml threshold=.5 device=$DEVICE $PRE_PROCESS inference-region=1 ! $AGGREGATE gv ...
```
we don't need this pipeline in Automated Self Checkout. Only the YOLOv8 pipeline.

antoniomtz commented on src/pipelines/age_recognition.sh:
```diff
@@ -0,0 +1,36 @@
+CLASS_MODEL_PATH="models/object_classification/age-gender-recognition-retail-0013/FP16-INT8/age-gender-recognition-retail-0013.xml"
+CLASS_MODEL_PROC_PATH="models/object_classification/age-gender-recognition-retail-0013/age-gender-recognition-retail-0013.json"
+
+gstLaunchCmd="gst-launch-1.0 $inputsrc ! $DECODE ! \
```
we don't need this pipeline in Automated Self Checkout. Only the YOLOv8 pipeline.

francis-tanyx replied:
removed

antoniomtz commented on smoke_test.sh:
```diff
@@ -89,12 +89,41 @@ waitForLogFile
+    teardown
+
+    # 5. Age pipeline: should see non-empty pipeline0.log contents
+    echo "Running Age Classification pipeline with logs..."
```
Remove this please, as we don't need Age/Gender use case here. Only YOLOv8

francis-tanyx replied:
removed

antoniomtz commented:
@francis-tanyx Made some comments

francis-tanyx commented:
> @francis-tanyx Made some comments

Thank you, will work with the team to follow up on the required changes

francis-tanyx commented on src/pipelines/yolov8s_roi.sh:
```diff
@@ -0,0 +1,42 @@
+ #!/bin/bash
+#
+# Copyright (C) 2024 Intel Corporation.
```
done

francis-tanyx commented on src/pipelines/people_detection.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+#
+# Copyright (C) 2024 Intel Corporation.
```
done

francis-tanyx commented:
> When running `make run-demo` I get an error - permission denied in the dlstreamer container src-OvmsClientGst-1 `/script/entrypoint.sh: line 137: /home/pipeline-server/pipelines/yolov8s_roi.sh: Permission denied`

added permission to yolov8s_roi.sh and also renamed to yolov8s.sh to reflect removal of ROI

ejlee3 reviewed (changes requested):
When running `YOLO=yolo8 make run-demo` I see that the OvmsClientGst container starts with the correct model, but the container does not keep running and no detection output is displayed.

NeethuES-intel commented:
@francis-tanyx please fix the merge conflicts

francis-tanyx commented:
@NeethuES-intel 

Apologies was busy the past week. Will look into the issues on @ejlee3's end first. 
but @ejlee3 can you try `YOLO=yolov8 make run-demo`?

francis-tanyx commented:
@NeethuES-intel conflict solved, had to rebase my git

francis-tanyx commented:
@ejlee3 We pushed a new fix, hopefully this fixes the issue.

ejlee3 reviewed (approved):
LGTM - worked for yolov8.

ejlee3 commented:
> @ejlee3 We pushed a new fix, hopefully this fixes the issue.

Thanks for fixing the issue @francis-tanyx I was able to run successfully on my system now.

francis-tanyx commented:
> > @ejlee3 We pushed a new fix, hopefully this fixes the issue.
> 
> Thanks for fixing the issue @francis-tanyx I was able to run successfully on my system now.

No problem, couldn't have done it without my Team's assistance.

antoniomtz commented on Makefile:
```diff
@@ -16,18 +17,46 @@ RESULTS_DIR ?= $(PWD)/results
+
+
+ifeq ($(YOLO),yolov8)
+download-models: download-yolov8s
```
No need to have a condition. We can download both all models when the user runs the download-models

antoniomtz commented on Makefile:
```diff
@@ -109,10 +138,10 @@ clean-telegraf:
 
 run-portainer:
-	docker compose -p portainer -f docker-compose-portainer.yml up -d
+	PIPELINE_SCRIPT=$(PIPELINE_SCRIPT) docker compose -p portainer -f docker-compose-portainer.yml up -d
```
There is no need to have PIPELINE_SCRIPT here

antoniomtz commented on Makefile:
```diff
@@ -109,10 +138,10 @@ clean-telegraf:
 
 down-portainer:
-	docker compose -p portainer -f docker-compose-portainer.yml down
+	PIPELINE_SCRIPT=$(PIPELINE_SCRIPT) docker compose -p portainer -f docker-compose-portainer.yml down
```
There is no need to have PIPELINE_SCRIPT here

antoniomtz commented on src/entrypoint.sh:
```diff
@@ -19,7 +19,6 @@ checkBatchSize() {
 }
 
 cid_count="${cid_count:=0}"
-CONTAINER_NAME="${CONTAINER_NAME:=gst}"
```
This may be needed for the performance_tools. That's how the performance scripts find the log files.

antoniomtz commented on src/entrypoint.sh:
```diff
@@ -122,18 +121,15 @@ fi
-touch /tmp/results/pipeline"$cid".log
-chown 1000:1000 /tmp/results/pipeline"$cid".log
-
+touch /tmp/results/r"$cid"_gst.jsonl
```
what's the reason for this change?
We need to test the performance tools and make sure this update doesn't break how the tools find and parse this log files.

antoniomtz commented on src/extensions/object_filter.py:
```diff
@@ -0,0 +1,114 @@
+'''
```
This logic belongs to loss-prevention. 
Automated selfcheckout provides only the basic CV pipelines.

antoniomtz commented on src/pipelines/yolov8s_roi.sh:
```diff
@@ -0,0 +1,38 @@
+ #!/bin/bash
```
ROI also belongs to loss prevention.

antoniomtz commented on src/pipelines/yolov8s.sh:
```diff
@@ -0,0 +1,38 @@
+echo "decode type $DECODE"
+echo "Run YOLOv8 pipeline on $DEVICE with batch size = $BATCH_SIZE"
+
+gstLaunchCmd="GST_DEBUG=1 GST_TRACERS=\"latency_tracer(flags=pipeline,interval=100)\" gst-launch-1.0 $inputsrc ! $DECODE ! gvadetect batch-size=$BATCH_SIZE model-instance-id=odmodel name=detection model=models/object_detection/yolov8s/FP32/yolov8s.xml device=$DEVICE $PRE_PROCESS inference-region=1 object-class="$CLASS_IDS" threshold=0.6 ! \
```
The idea is to update the existing yolov5 pipelines and call it yolo.sh , where using an argument can change the value of model.

This way we can reuse the same bash script for both models.

antoniomtz reviewed (changes requested):
@francs

antoniomtz commented:
@francis-tanyx thank you for your contributions. Based on my feedback, this PR needs a major refactor which I can work on that. 
Can you please push this changes into a new branch other than `main` ? this way I can have access to your contributions and make the updates I suggested.

francis-tanyx commented:
> @francis-tanyx thank you for your contributions. Based on my feedback, this PR needs a major refactor which I can work on that. Can you please push this changes into a new branch other than `main` ? this way I can have access to your contributions and make the updates I suggested.

Okay noted. Any needs on the new branch naming required?

francis-tanyx commented:
@antoniomtz hi I just wanted to follow up on the branching, it seems I dun have access to create a new branch. Would it be possible if you could create the branch?
